### ISSUE TITLE
First release TODO

### ISSUE THREAD

**Issue by knuxify** [aspinwall-ui/aspinwall-launcher#1]
This issue outlines the work that has to be done before the first release can be finished.

## Launcher

  - [x] Configuration file
    * [x] Consider moving it to GSettings? (0ef3a712975e99fb9b3942d2c5e0cae097cd058d)
    * [x] Add option for changing how the date and time are displayed (280f8806b277c0a28df2acf714b3947fcc6d643c)
  - [x] Widget management
    * [x] Adding widgets
    * [x] Removing widgets
    * [x] Moving widgets (2ddddebb2b41e3228ae9b9f1b900882756f3adbd)
    * [x] Refreshing widgets (+ autorefresh) (545acc99db9fab76695c005338d38d866a985c57)
    * [X] Widget configuration menu (per-widget) (4c8fc655db2a72b1df78a7b2c9e854223c8a1e59)
    * [X] Widget management mode (6e4afabfce3affe7947c10bfa596b4c6c0d43b6a)
  - [X] App drawer (58c4063ddfe5989798b0187fcc262323af794400)
     - Add categories (probably won't bother with this)
     - [X] Add favorites (ee222040862bd96483b68041e51507929f02940e)
  - [X] Desktop backgrounds (9cd2207ced4a2760ad0016418c5d79c357c3ef82)
     - [X] Dimming in widget editing mode
     - [X] Wallpaper preview in settings (3efc26b54ab79ecd7c08b58db648156cc3f664c6)
     - [X] Slideshows (5805ba3f6a68559803370b07c987679899114546)
       - [X] Drag-and-drop for wallpapers in settings (b36ffdd49bd6ac25dfc3a2ccad427d72deb595e9)
     - [X] Wire up other fit options (37c056e374dacd227b798c8ccf1ee84e7e3ee64c)
     - [X] Select available wallpaper by default (currently it hard-codes a GNOME wallpaper) (93cbfa9aef652d81ccea5a72704473a5cfcec0d6)
  - [X] Dark mode (f6a06d6c171c3558e75240e78528cbccac180a19)
  - [X] Switch to AdwCarousel on vertical screens (0bad3c473fa22b5f673b6f12f8c3ce7c0a15d82c, had to use AdwLeaflet; might need some extra gesture tweaks)
  - [x] Rewrite "Add widgets" dialog to use listviews (a ListBox might work fine too) (414ef6981370bf09615fd305ac1602fedac42232)
    - [X] Might consider taking a look at actual listviews though, unless search can be done in a listbox (moved to listview, and added search: 526492a7ce82c8af7db5b762981be4567a2381b3)
  - [x] Global custom color shorthands instead of relying on the system theme (ffb461c60cb4a8fcd3b969ffcbb51ce62239376d, fe9cf6a87c2f8cba1c1f9d31065d05d06c93cd10)
  - [X] Only allow one launcher window to be open at a time (c77dfe1623fbf2cfa370133cc2af6238bea82834)
  - [X] Launcher settings (e92a8a28a4a125c32b7209ec895d158c6270a43d)
  - [X] About Aspinwall menu (5c9f049bf50cd319cba2b3734a72e235e31a3259)
  - [X] Hide favorites in the app chooser when there are none (f4402e44090120d310c59c78f81e770dc37a9221)
  - [X] Status pages for empty search results (app chooser, widget chooser) (38122ce75857556142daa99e2f62294106b6bb98, a84a6c6a323475a8bc1e346e451ff6fe85e985c3)
  - [X] "Set to default" option for time format settings (16002112b8f6f0c24d5d71e42ffcc72f342c8d2e)

### Code quality

  - [X] Make AdwWidget use templates (we're not inheriting from it anymore) (fca089a255941e9c044fef80e39b8e3d27d39091)
  - [X] Figure out if it's possible to viably move factories/models to UI files (IMO this is not needed, keeping it in-code makes it cleaner + easier to use with our custom objects (in the case of widgets for the widget chooser, for example))
  - [X] Add tests (df278f52e00e3c6675f8591a23b3041e95b61fb7, failed back then but works now that we switched to Debian in ac59e5666863d2dd6522c07261e7d5e5cf6bba29)

### Bugfixes

  - [X] Fix memory leak on app chooser menu re-opening (look into profiling tools?) (243b5c717c7bb2ebb29817d8f23facc8bdb53a0f)
  - [X] Fix wallpaper preview memory leak (ec8820d1b6c9fcb831c39cff9be3cdbd0d12cfdc)
  - [X] Fix crash when exiting search mode in preferences (806ebb7221f921c739f69d8fa85baa3b10b180ee)
  - [X] Change background when it's removed from available wallpapers (cb2df7602549e6d5d4fbf2d8e590132992cb2d9d)
  - [X] Settings not appearing on first launch? Might be related to the lack of a default wallpaper, though. (Can confirm this no longer happens.)
  - [ ] Fix drag-and-drop on touchscreens (might be a GTK issue?)
  - [ ] Fix memory leak when searching for apps (flowbox issue; I'll try to investigate this if I find the time, otherwise it'll have to be postponed.)

## Meta

 - [X] Add translations (882c5ba5709f3d3c93ac4713c34918da22bc8a38)
 - [X] Add meson.build file (170823613a995ee7850d75b2d2afec3324867d3e)
 - [X] Development documentation (this is a pretty broad topic; I'm ticking this off as done, although the docs can definitely still be improved upon.)
 - [X] Add documentation for translating widgets (9bcfb75d5563bc7034cdca27f1297b0750a310be)
 - [X] Create example widget repo (66657ee8f249b4aa6aab19094dcb13e414143f3d, https://github.com/aspinwall-ui/aspinwall-example-widget)

Once the first release is finished:
 - Package for Alpine Linux/postmarketOS
 - Package for Manjaro Mobile/Arch
 - Package for Mobian

# Postponed for next release
  - OLED mode (black background, white text); I think what I'm going at here is more of a "pure black" theme, which could be done, but would require extra styling. As for actual OLED features and general convenience features like burn-in prevention, etc - slideshow wallpapers should help out with this.
  - Proper support for inheriting backgrounds from GNOME - currently it only works for simple images, but fails on .xml files ("slideshow" wallpapers), and doesn't respect light/dark settings. Might end up removing support for these wallpapers entirely, though.
  - (ERRATA) Add custom translation loader workaround for Alpine (might solve this one on Alpine's side rather than trying to hack it into Aspinwall... seems like what we'll need is a separate locale library that can access system settings, and prefferably replace the `_locale` module in Python without having to adapt code)
  - Clean up code for the launcher settings (there should be a way to bind the switches directly to the config options in some cases, in others we can make reusable template functions)
  - Improve tests, maybe? Currently their main purpose is to quickly check if the launcher even works or not, which is pretty handy in figuring out if new commits don't have obvious issues, and makes packaging far easier. It would be nice to have more comprehensive tests though, especially for the launcher settings and shell, and various edge cases + things that we normally wouldn't be able to test otherwise.
  - Use listboxes or other list widgets for the widgetbox; our current solution is a bit hacky, although it works relatively well. My main concern is lack of scrolling on widget drag-and-drop, as well as potential issues with keyboard support.
  - Show widget editing button on hover
  - Focused/unfocused mode (show widget add button when the window is active, hide after ~1min of inactivity; during inactivity lower widget opacity)
  - Clock size toggle
  - Make a logo... or find someone who can
  - Accessibility improvements

**Comment by knuxify** [comment]
For translations: I'm thinking about adding the project to [Hosted Weblate](https://hosted.weblate.org), but their rules require that added projects must have been in active development for at least 3 months, and since this project is still younger than that I need to wait.

Until then, here's what would need to be added to the glossary:

- Widget (for reference, see e.g. Android's widgets or KDE Plasma's widgets)
- App chooser (can be translated as "app choice menu" if needed)
- Widget chooser (can be translated as "widget choice menu" if needed)
- Launcher (usually this is just "launcher", independent of language - but perhaps some languages have alternatives?)
- Control center/control panel (still need to decide on wording)
- Clock box (can be translated as "clock container")
- Widget box (can be translated as "widget container")
- Common dialog box button commands (Cancel, Select, Ok, etc.)
- Wallpaper, wallpaper style
- Status bar/panel

**Comment by knuxify** [comment]
Moving the shell-specific TODOs here:

## Shell

Currently the biggest hurdle is figuring out how to tie everything together; there have been a few things that were added to achieve this:

 * Currently each shell component is built around a custom Surface class which we'll be able to expand as we please (i.e. with both X11 and Wayland-specific stuff).
 * We will still need a window manager for the whole thing. My current idea is to use [herbstluftwm](https://herbstluftwm.org/) (or a fork of it, if it turns out to be necessary), and that's for a few reasons:
   - It's pretty easy to control; the configs are basically shell scripts, and commands are sent through a CLI utility. ~~If it turns out to be controllable from Python directly (outside of using os.exec), it would be great.~~ yup, it has python bindings!
   - From what I recall from using it, its window sizing logic is fairly aggressive, and tries to resize windows to fit at all costs. This might cause some issues, but it certainly makes resizing windows to fit mobile screens easier.
   - It's a tiling window manager, which works out for us.
   - It uses X11, so it will run on pretty much everything.

 - [X] Panel
   - [X] Clock (2ca4cb207e0cd87bf44157f4ead47cc58095cfc1)
   - [X] Status icons (b7782fa8f03d188d55a8274894b5f6de09633637, 91e62cc4da4bb985e727b31314226c91c9b6cc9d)
     - [X] Wi-Fi/cellular status (4f3f679b1ad5bd0707801534b1593750111db2ba)
     - [X] Togglable battery percentage display (this is already coded in, just needs a config option) (49cb538194bee359aa6d2a729052a7be700d9612)
 - [X] Control center (shows up after dragging panel) (f8799343083dfdfcd2233803a5158f7ab0572916)
     - [X] Settings buttons (basic code in 723f7fc92c51b3eea814db2a60c6f870e887be81, more to follow)
     - [X] Fix initial animation (fixed this while working on 723f7fc92c51b3eea814db2a60c6f870e887be81)
     - [ ] Power button
     - [ ] Do something with the "settings" button, maybe? I'll probably just make it insensitive.
 - [X] Notification handler (97ba87ca9c04bd2f58c9dc7cc31fe89da1b6fdbb, also added to the control panel)
   - [X] Actions (a927054130f963906f4bde9f5c4bd0beddec135b)
   - [X] Pin urgent messages to top (b5b0d64f4844eb91d05e29c07dc3d28c92835895)
   - [X] Icon data support (c87b3bcb81bbed505dac69ce7794a83a06d38ba9)
   - [X] Progress bar support (be9ae2268226f287de94f9e3064dd84cb1928644)
 - [X] Notification pop-ups (65381d2f7c0a9dd4304a6ae802ad9ea139534535)
    - [ ] Don't show when the control panel is open
 - Settings backends for:
     - [X] Wi-Fi (networkmanager; 4f3f679b1ad5bd0707801534b1593750111db2ba, but still rather unstable. Ticking off so that I can focus on other tasks.)
     - [X] Volume (pulseaudio; b1a7ecd2cf79ac4d4bbd70f8337342abad75e23d)
 - [X] Window switcher (fcef74faa3b58efc16717aea58eb881477ae684f, 480f2644a04bbdb2c89554bceda72d2bb49c6899)
    - [X] Window list refreshing (23823957032e2236a8096ae06392a04df6dcda7d)
    - [ ] Window positioning (herbstluftwm?) 
 - [X] Window moving for shell surfaces (51235b027f55cd26d937437c74dcd373cb04ca0f, 481c60a9956cceabb9470cc55b8d0df4d10cf895)
 - [ ] Fix notification popup not appearing... we might have to set it as a custom floating window? Not 100% sure what this will involve.

## Postponed

- Wayland support - the initial release will run on Xorg, so as to support as much devices as possible - I'll investigate a Wayland solution, which might involve patching wlroots to run on devices with fbdev framebuffers... or just keeping Xorg here and adding Wayland as an extra option for those who can run it.
- Settings backend for cellular (modemmanager/ofono?)
- Bluetooth support (bluez), unless I manage to fit it in before the deadline and/or find an easy to use library for it
- Support for screen rotation; in general, due to the way we handle sizing, everything will break if the screen resolution changes. Also, I have yet to investigate potential APIs for controlling screen rotation.
- Screen brightness settings
- Proper on-screen keyboard support

**Comment by knuxify** [comment]
## Widgets
   * [x] Welcome widget
   * [ ] Weather
   * [ ] To-do list
   * [ ] World clock
   * [ ] Fun fact/fortune
   * [ ] Quick shortcuts/app launcher

**Comment by knuxify** [comment]
This pretty much turned into the first pre-release TODO (to make it before yh4f's March 31st deadline...), and given the rather large amount of finished points here and the general chaos in the "Postponed" sections, I'd rather move the things we still have to do into a separate issue.

See #3.
